### ISSUE TITLE
URL misinterpretation with quotes and parentheses

### ISSUE THREAD

**Issue by vlakoff** [notepad-plus-plus/notepad-plus-plus#15941]
Even after #15058 has landed, I encountered a case where the last quote is still erroneously in the link:
```
('http://example.com')
```

The issue is caused by the closing parenthese (which is not captured in the link).

Additional observations:
* Same problem with a closing square bracket: `['http://example.com']`
* No problem if there is a space before the closing parenthese (or square bracket): `( 'http://example.com' )`
* No problem with a curly bracket: `{'http://example.com'}`
* To clarify: the opening parenthese/bracket/etc have no effect on the issue.

Considering that the closing parenthese is not captured in the link, the URL detection mechanism is already able to figure out that it is not part of the URL. Thus, I guess that, using the same "quote elimination mechanism", we should also be able to handle the case I reported here.

ping: @MarkusBodensee

**Comment by MarkusBodensee** [comment]
Thank you for pinging, @vlakoff

Yes, it is an issue with the closing parenthese/bracket and how it is handled. Without space before, the closing parenthese is considered as a valid part of the URL. The closing parenthese is trimmed from the url after the apostrophes are handled. So switching the order could be a solution.

Another possibility could be, to look for an opening parenthese before the url (I already wrote about it in the PR). But there could be cases like, we also need to consider:
```
Some Text (Example: https://www.example.com/)
Some Text (Example: 'https://www.example.com/')
```

I will look into this issue and do some testing.

**Comment by vlakoff** [comment]
Just a thought: another solution could be to remove "balanced" characters around the string.
As an example, `(''http://example.com/'')` would be trimmed to `http://example.com/`.

For reference, here is an AI-generated code:

<details>
  <summary>Trimming balanced characters from a string in C++</summary>

```cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <stack>

std::string trimBalancedCharacters(const std::string& str) {
    if (str.empty()) return str;

    std::unordered_map<char, char> matchingPairs = {
        {'(', ')'}, {'[', ']'}, {'\'', '\''}
    };

    std::stack<char> charStack;
    size_t start = 0, end = str.length() - 1;

    // Check from the beginning for balanced opening characters
    while (start < str.length() && matchingPairs.find(str[start]) != matchingPairs.end()) {
        charStack.push(str[start]);
        start++;
    }

    // Check from the end for matching closing characters
    while (end > start && !charStack.empty() && matchingPairs[charStack.top()] == str[end]) {
        charStack.pop();
        end--;
    }

    if (charStack.empty()) {
        return str.substr(start, end - start + 1);
    }

    return str;
}

int main() {
    std::string str = "\'(example)\'";
    std::cout << "Original: " << str << "\n";
    std::cout << "Trimmed: " << trimBalancedCharacters(str) << "\n";
    return 0;
}
```
</detail>

**Comment by MarkusBodensee** [comment]
Yes, the "remove balanced characters" solution was also what I was thinking about above (and also while writing in the PR). But since it would break the examples which I posted above, it is not a good solution.

I think, a handling for balanced paranthesis should be implemented, but balanced paranthesis must be handled different than balanced apostrophes.

**Comment by vlakoff** [comment]
A quick bump on this issue — I've noticed the problem also occurs when the quote is followed by a comma.

Example:

```php
$urls = [
    'https://example.org/foo', // closing quote is erroneously included in the hyperlink
    'https://example.org/bar'  // closing quote is correctly excluded from the hyperlink
];
```

(By the way, a simple enhancement to the feature from #15058 would be to add support for double quotes, in addition to single quotes and backticks.)
